### PR TITLE
Add support for custom padding in DNN preprocessing

### PR DESCRIPTION
**Pull request by Abdurrahheem**
This PR add functionality for specifying value in padding.
 It is required in many preprocessing pipelines in DNNs such as Yolox object detection model

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 1]
```diff
@@ -1212,7 +1212,7 @@ CV__DNN_INLINE_NS_BEGIN
         CV_WRAP Image2BlobParams();
         CV_WRAP Image2BlobParams(const Scalar& scalefactor, const Size& size = Size(), const Scalar& mean = Scalar(),
                             bool swapRB = false, int ddepth = CV_32F, DataLayout datalayout = DNN_LAYOUT_NCHW,
-                            ImagePaddingMode mode = DNN_PMODE_NULL);
+                            ImagePaddingMode mode = DNN_PMODE_NULL, Scalar fillvalue = 0.0);
```
Comment (asmorkalov) [inline modules/dnn/include/opencv2/dnn/dnn.hpp:1215]: `borderValue` is better. It's consistent with other API functions.

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 2]
```diff
@@ -1221,6 +1221,7 @@ CV__DNN_INLINE_NS_BEGIN
         CV_PROP_RW int ddepth;   //!< Depth of output blob. Choose CV_32F or CV_8U.
         CV_PROP_RW DataLayout datalayout; //!< Order of output dimensions. Choose DNN_LAYOUT_NCHW or DNN_LAYOUT_NHWC.
         CV_PROP_RW ImagePaddingMode paddingmode;   //!< Image padding mode. @see ImagePaddingMode.
+        CV_PROP_RW Scalar paddingmodefillvalue;   //!< Value used in padding mode for padding.
```
Comment (asmorkalov) [inline modules/dnn/include/opencv2/dnn/dnn.hpp:1224]: `borderValue`

File: modules/dnn/src/dnn_utils.cpp [hunk 3]
```diff
@@ -196,7 +196,7 @@ void blobFromImagesWithParamsImpl(InputArrayOfArrays images_, Tmat& blob_, const
                 int bottom = size.height - top - rh;
                 int left = (size.width - rw)/2;
                 int right = size.width - left - rw;
-                copyMakeBorder(images[i], images[i], top, bottom, left, right, BORDER_CONSTANT);
```
Comment (asmorkalov) [inline modules/dnn/src/dnn_utils.cpp:199]: I propose to get rid of copy and use ROI.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Review by fengyuentau — COMMENTED** [review]
Does YOLO family use different padding values?

**Comment by Abdurrahheem** [comment]
> Does YOLO family use different padding values?

It uses 114. We do not have such support out of the box

**Comment by dkurt** [comment]
> It uses 114. We do not have such support out of the box

It seems like not a padding but mean value. So after subtraction of mean 114 final padding is zero

**Comment by fengyuentau** [comment]
> It uses 114. We do not have such support out of the box.

Ok. More questions:

1. 114 is needed by which YOLO version?
2. Do other YOLO versions use different value?

---

> It seems like not a padding but mean value. So after subtraction of mean 114 final padding is zero

IIRC, the letterbox tranformation pads the image to a certain scale, for example, h:w=1:1; then when it is resized, it does not make distortion on the potential objects.

**Comment by LaurentBerger** [comment]
> It uses 114. We do not have such support out of the box

Agree I think values are necessary like pytorch https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html#torch.nn.functional.pad or tensorflow https://www.tensorflow.org/api_docs/python/tf/pad

File: modules/dnn/include/opencv2/dnn/dnn.hpp [hunk 4]
```diff
@@ -1221,7 +1221,7 @@ CV__DNN_INLINE_NS_BEGIN
         CV_PROP_RW int ddepth;   //!< Depth of output blob. Choose CV_32F or CV_8U.
         CV_PROP_RW DataLayout datalayout; //!< Order of output dimensions. Choose DNN_LAYOUT_NCHW or DNN_LAYOUT_NHWC.
         CV_PROP_RW ImagePaddingMode paddingmode;   //!< Image padding mode. @see ImagePaddingMode.
-        CV_PROP_RW Scalar paddingmodefillvalue;   //!< Value used in padding mode for padding.
+        CV_PROP_RW Scalar paddingmodeborderValue;   //!< Value used in padding mode for padding.
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/dnn.hpp:1224]: Not sure that this is a good naming for a public field (as the other names too).

File: modules/dnn/test/test_misc.cpp [hunk 5]
```diff
@@ -93,6 +93,42 @@ TEST(blobFromImageWithParams_4ch, NHWC_scalar_scale)
     }
 }
 
+TEST(blobFromImageWithParams_CustomPadding, letter_box)
+{
+    Mat img(40, 20, CV_8UC4, Scalar(2, 2, 2, 2));
+
+    // Custom padding value that you have added
+    Scalar customPaddingValue(1, 1, 1, 1); // Example padding value
```
Comment (asmorkalov) [inline modules/dnn/test/test_misc.cpp:101]: It makes sense to use different value for different channels and check each channel value.

File: modules/dnn/test/test_misc.cpp [hunk 6]
```diff
@@ -93,6 +93,42 @@ TEST(blobFromImageWithParams_4ch, NHWC_scalar_scale)
     }
 }
 
+TEST(blobFromImageWithParams_CustomPadding, letter_box)
+{
+    Mat img(40, 20, CV_8UC4, Scalar(2, 2, 2, 2));
```
Comment (asmorkalov) [inline modules/dnn/test/test_misc.cpp:98]: It makes sense to use different value for different channels and check each channel value.

File: modules/dnn/test/test_misc.cpp [hunk 7]
```diff
@@ -93,6 +93,42 @@ TEST(blobFromImageWithParams_4ch, NHWC_scalar_scale)
...
+    param.paddingmode = DNN_PMODE_LETTERBOX;
+    param.borderValue = customPaddingValue; // Use your new feature here
+
+    // Create blob with custom padding
+    Mat blob = dnn::blobFromImageWithParams(img, param);
+
+    // Create target blob for comparison
+    Mat targetBlob = dnn::blobFromImage(targetImg, 1.0, targetSize);
+
+    EXPECT_EQ(0, cvtest::norm(targetBlob, blob, NORM_INF));
```
Comment (dkurt) [inline modules/dnn/test/test_misc.cpp:129]: Where is the check for `customPaddingValue`?

Comment (Abdurrahheem) [reply modules/dnn/test/test_misc.cpp:129]: What do you mean? Be more specific please.

Comment (dkurt) [reply modules/dnn/test/test_misc.cpp:129]: There is no connection of `customPaddingValue` variable and the rest of test. If I change the value - test fails. I expected something like that:
```cpp
TEST(blobFromImageWithParams_CustomPadding, letter_box)
{
    Mat img(40, 20, CV_8UC4, Scalar(0, 1, 2, 3));

    // Custom padding value that you have added
    Scalar customPaddingValue(0, 1, 2, 3); // Example padding value

    Size targetSize(20, 20);

    Mat targetImg = img.clone();
    cv::copyMakeBorder(targetImg, targetImg, 0, 0, 0, targetSize.width, BORDER_CONSTANT, customPaddingValue);

    // Set up Image2BlobParams with your new functionality
    Image2BlobParams param;
    param.size = targetSize;
    param.paddingmode = DNN_PMODE_LETTERBOX;
    param.borderValue = customPaddingValue; // Use your new feature here

    // Create blob with custom padding
    Mat blob = dnn::blobFromImageWithParams(img, param);

    // Create target blob for comparison
    Mat targetBlob = dnn::blobFromImage(targetImg, 1.0, targetSize);

    EXPECT_EQ(0, cvtest::norm(targetBlob, blob, NORM_INF));
}
```

Comment (dkurt) [reply modules/dnn/test/test_misc.cpp:129]: There is no connection of `customPaddingValue` with the rest of test. If it's changed - test fails. So I recommend something like that:
```cpp
TEST(blobFromImageWithParams_CustomPadding, letter_box)
{
    Mat img(40, 20, CV_8UC4, Scalar(0, 1, 2, 3));

    // Custom padding value that you have added
    Scalar customPaddingValue(0, 1, 2, 3); // Example padding value

    Size targetSize(20, 20);

    Mat targetImg = img.clone();
    cv::copyMakeBorder(targetImg, targetImg, 0, 0, targetSize.width / 2, targetSize.width / 2, BORDER_CONSTANT, customPaddingValue);

    // Set up Image2BlobParams with your new functionality
    Image2BlobParams param;
    param.size = targetSize;
    param.paddingmode = DNN_PMODE_LETTERBOX;
    param.borderValue = customPaddingValue; // Use your new feature here

    // Create blob with custom padding
    Mat blob = dnn::blobFromImageWithParams(img, param);

    // Create target blob for comparison
    Mat targetBlob = dnn::blobFromImage(targetImg, 1.0, targetSize);

    EXPECT_EQ(0, cvtest::norm(targetBlob, blob, NORM_INF));
}
```

**Review by dkurt — APPROVED** [review]

**Review by fengyuentau — APPROVED** [review]
👍

File: modules/dnn/src/dnn_utils.cpp [hunk 8]
```diff
@@ -196,7 +196,7 @@ void blobFromImagesWithParamsImpl(InputArrayOfArrays images_, Tmat& blob_, const
...
+                copyMakeBorder(images[i], images[i], top, bottom, left, right, BORDER_CONSTANT, param.borderValue);
```
Comment (opencv-alalek) [inline modules/dnn/src/dnn_utils.cpp:199]: > `images[i], images[i]`

Operation is not in-place.
We should not use the same argument for `src` and `dst` - this is anti-pattern and bad practice.
